src/authentication/authentication-hapi-plugin.ts

```diff
@@ -95,20 +85,52 @@ class AuthenticationHapiPlugin extends HapiPlugin {
   public async getTeamHandler(request: Hapi.Request, reply: Hapi.IReply) {
     try {
       const credentials = request.auth.credentials as AccessToken;
-      const user = await this.gitlab.getUserByEmailOrUsername(sanitizeUsername(credentials.sub));
-      const teams = await this.gitlab.getUserGroups(user.id);
+      const teams = await this.gitlab.getUserGroups(sanitizeUsername(credentials.sub));
       return reply(teams[0]); // NOTE: we only support a single team for now
     } catch (error) {
       this.logger.error(`Can't fetch user or team`, error);
       return reply(Boom.wrap(error, 404));
     }
   }
 
+  /**
+   * Allows fetching a team-token for a team that:
+   *  1. An authenticated user belongs to
+   *  2. An admin user has specified in the request by a team's id or name
+   */
   public async getTeamTokenHandler(request: Hapi.Request, reply: Hapi.IReply) {
     try {
-      const teamId = parseInt(request.paramsArray[0], 10);
-      const teamToken: TeamToken[] = await teamTokenQuery(this.db, undefined, teamId);
-      if (!teamToken.length) {
+      const credentials = request.auth.credentials as AccessToken;
+      const userName = sanitizeUsername(credentials.sub);
+      const isAdmin = await this.isAdmin(userName);
+      const userTeams = await this.gitlab.getUserGroups(userName);
+      const teamIdOrName = request.paramsArray.length ? request.paramsArray[0] : undefined;
+
+      let requestedOwnTeam: Group | undefined;
+      if (userTeams && teamIdOrName) {
+        requestedOwnTeam = userTeams.find(findTeamByIdOrName(teamIdOrName));
+      }
+
+      let teamId: number | undefined;
+
+      if (requestedOwnTeam) { // a request for a team that the user belongs to
+        teamId = requestedOwnTeam.id;
+      } else if (isAdmin && teamIdOrName) { // An admin can get any team's token
+        teamId = await this.teamIdOrNameToTeamId(teamIdOrName);
```

Comment:
I can't see this. If the user is an admin requesting a token for a specific team that she doesn't belong to, then `const userTeams = await this.gitlab.getUserGroups(userName);` is being called only if the the specific team was specified by name.